File: Java/src/main/java/edu/bu/benchmarks/DataSerialization.java
package edu.bu.benchmarks;

import com.google.gson.Gson;
import edu.bu.filehandler.FileHandler;
import edu.bu.filehandler.LogFileHandler;
import edu.bu.filehandler.ReadTweet;
import edu.bu.tweet.TweetStatus;
import org.apache.log4j.PropertyConfigurator;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Properties;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.stream.Stream;

public class DataSerialization {

    public static void main(String[] args) throws IOException {

        Properties prop = new Properties();

        PropertyConfigurator.configure(prop);
        prop.setProperty("log4j.appender.WORKLOG.File", "log4j.properties");

        PropertyConfigurator.configure("log4j.properties");

        // Before any experiments create a hot gc
        // Create a hot garbage collector
         //Garbage.activateGarbageCollector();
//
//        System.out.println("Activating Java garbage collector!");
      //  System.gc();
//        System.out.println("Java garbage collector is made hot!");

        // Set input file path:
        String inputFile = args[0];

        int serializationType = Integer.parseInt(args[1]);

        //Set output file path
        String outFile = args[2];

        // set number of tweets:
        long numberOfTweets=Long.parseLong(args[3]);

        int round=Integer.parseInt(args[4]);

        boolean taskset=Boolean.parseBoolean(args[5]);

        String logFileName="bin/benchmark/writeobjects/result_java_writeobjects_" + numberOfTweets + "_" + round + ".txt";


        FileHandler fileHandler = new FileHandler(outFile, serializationType);
        fileHandler.prepareToWrite();

        BlockingQueue<String> queue = new ArrayBlockingQueue<>(1000);
        ReadTweet readTweet = new ReadTweet(queue,inputFile);
       Thread rt=  new Thread(readTweet);
       rt.start();
        //starting consumer to consume raw tweet from queue

        double elapsedSeconds=0;

        try{
            String line;
            readTweet.setReadStatus(true);
            for (long i=0;i<numberOfTweets;i++){
                line = queue.take();
                Gson gson = new Gson();
                TweetStatus tweetStatus = gson.fromJson(line, TweetStatus.class);
                long tmpTime = System.nanoTime();
               // fileHandler.appendObjectToFile(tweetStatus);
                elapsedSeconds+= (System.nanoTime() - tmpTime);
            }
            readTweet.setReadStatus(false);
            queue.take();
        }catch(Exception e) {
            e.printStackTrace();
        }
        fileHandler.appendObjectToFileFlush();

        // Time Calculation
        elapsedSeconds /= 1000000000.0;

        LogFileHandler logFileHandler=new LogFileHandler(logFileName);
        logFileHandler.addLog(false,serializationType,true,"TweetStatus",fileHandler.getIoTime(),elapsedSeconds,taskset);


    }
}

File: Rust/src/benchmarks/DataSerialization.rs
use std::fs::File;
use std::io::{BufReader, BufRead};
use std::io;
use crate::filehandler::FileHandler::FileHandler;
use std::time::{Duration, Instant};
use crate::tweetStructs::TweetStatus::TweetStatus;
use serde::{Serialize, Deserialize};
use crate::filehandler::LogFileHandler::LogFileHandler;


//pub fn DataSerialization(inputFile:String, serializationType:i32, outFile:String, numberOfTweets:i64, round:i32, taskset:bool){
//write  $datapath $serialization_type $outpath $numberOfTweets $r false
pub fn DataSerialization(args:Vec<String> ) -> io::Result<()>{

    let mut inputFile:String = args[2].to_string();

    let serializationType:i32 = args[3].parse().unwrap();

    //Set output file path
    let mut outFile:String = args[4].to_string();

    // set number of tweets:
    let  numberOfTweets:i64=args[5].parse().unwrap();

    let  round:i32=args[6].parse().unwrap();

    let  taskset:bool= args[7].parse().unwrap();

    let log_file_name:String = format!("{}{}{}{}{}", "bin/benchmark/writeobjects/result_rust_writeobjects_", numberOfTweets, "_", round, ".txt");
    let logfile=LogFileHandler::new(String::from(log_file_name));

    let mut file_handler = FileHandler::new(outFile, serializationType);
    file_handler.prepareToWrite();

    let f = File::open(inputFile)?;
    let f = BufReader::new(f);

    let mut elapsedSeconds:Duration = Duration::new(0, 0);
      for line in f.lines() {

         let  tweetStatus:TweetStatus = serde_json::from_str(&line.unwrap()).unwrap();

          let start = Instant::now();
          file_handler.appendObjectToFile(tweetStatus);
          let duration = start.elapsed();
          elapsedSeconds+=duration;
      }

    // Finalize and close all open files:
    file_handler.appendObjectToFileFlush();

    logfile.add_log(false, serializationType, true, String::from("TweetStatus"), elapsedSeconds, elapsedSeconds, taskset);

    Ok(())
}